src/widgets/analytics/analytics.ts

```diff
@@ -0,0 +1,202 @@
+import { createDocumentationMessageGenerator } from '../../lib/utils';
+import { SearchParameters, SearchResults, Widget } from '../../types';
+
+type AnalyticsWidgetParams = {
+  pushFunction(
+    formattedParameters: string,
+    state: SearchParameters,
+    results: SearchResults
+  ): void;
+  /**
+   * @default 3000
+   */
+  delay: number;
+  /**
+   * @default false
+   */
+  triggerOnUIInteraction: boolean;
+  /**
+   * @default false
+   */
+  pushInitialSearch: boolean;
+  /**
+   * @default false
+   */
+  pushPagination: boolean;
+};
+
+const withUsage = createDocumentationMessageGenerator({ name: 'analytics' });
+
+function analytics(
+  {
+    pushFunction,
+    delay = 3000,
+    triggerOnUIInteraction = false,
+    pushInitialSearch = true,
+    pushPagination = false,
+  }: AnalyticsWidgetParams = {} as AnalyticsWidgetParams
+): Widget {
+  if (!pushFunction) {
+    throw new Error(withUsage('The `pushFunction` option is required.'));
+  }
+
+  type AnalyticsState = {
+    results: SearchResults;
+    state: SearchParameters;
+  } | null;
+
+  let cachedState: AnalyticsState = null;
+
+  const serializeRefinements = function(parameters: object): string {
+    const refinements: string[] = [];
+
+    for (const parameter in parameters) {
+      if (parameters.hasOwnProperty(parameter)) {
+        const values = parameters[parameter].join('+');
+
+        refinements.push(
+          `${encodeURIComponent(parameter)}=${encodeURIComponent(
+            parameter
+          )}_${encodeURIComponent(values)}`
+        );
+      }
+    }
+
+    return refinements.join('&');
+  };
+
+  const serializeNumericRefinements = function(
+    numericRefinements: any
```

Comment:
To be more specific, it could be typed as `QueryParameters['numericRefinements']` (from `algoliasearch-helper`)